### ISSUE TITLE
context support

### ISSUE THREAD

**Issue by bmcmahen** [bmcmahen/toasted-notes#5]
Given the current API, jsx in render props or elements won't get access to context. Explore ways to support context while maintaining the current API, and inform users of ways to supply context to their render callback (wrap it with context providers).

**Comment by georgefeast** [comment]
I think I ran into this problem today.. Tried to replicate how your toasts work in Sancho UI with my own render callback component + toasted-notes where <ANotif/> is a custom component for displaying the notification.

`  toaster.notify(
    ({ onClose, id }) => (
      <Fragment>
        <GlobalStyle />
        <div>
          <ANotif
            intent={intent}
            message={message}
            handleNotifClosed={onClose}
          />
        </div>
      </Fragment>
    ),
    options
  )`

It works as long as ANotif contains just a message with a div but breaks as soon as I add styled-components inside. Is this what you're referring to in this issue / do you know any workaround? Sancho and toasted-notes are great by the way!

**Comment by bmcmahen** [comment]
Hmm yeah, it'll break if you rely on theme variables with styled components. The key is to wrap your render callback with your theme, like so:

```jsx
function CustomAlert() {
  <div>Custom</div>
}

const CustomAlertWithTheme = withTheme(CustomAlert)

toast.notify(() => <CustomAlertWithTheme />
```

I think one possible solution on our end (and something I've been a bit reluctant to do) is to provide a `ToastedNotesContainer` component which provides a target render element inside of your React component tree, thus ensuring that theme context is passed to the render callback.

**Comment by georgefeast** [comment]
Ah okay yeah I was also getting this error 

![Screen Shot 2019-04-03 at 8 54 43 AM](https://user-images.githubusercontent.com/2541627/55450873-270a1500-562d-11e9-9cae-18b27b976182.png)
 
Is this related to the same issue? Am I meant to set container somehow or pass the ref through?

**Comment by bmcmahen** [comment]
Huh, that's odd. I don't think this is related. It's indicating that the ref isn't being set for some reason and I'm not sure why that'd be the case. The ref should be referring to the `<animated.div />` container element which should definitely be there. Check [here](https://github.com/bmcmahen/toasted-notes/blob/7d3ccc856c47db8997ba3313d9897fae1c5eee2f/src/Alert/Message.tsx#L148).

**Comment by pashazz** [comment]
Hello! Bumping this thing up.

I wasn't able to use context properly. I want to use Apollo client inside a toast (this is GraphQL thing) so here's what I'm doing:
```
export const taskComponent = (ref: string): React.ComponentType<{}> => {
  return withApollo(({client}: ClientProps) => {
    return <ApolloProvider client={client}>
      <Subscription subscription={TaskInfoUpdateDocument}
                    variables={{ref: ref}}>
        {({data}) => {
          return (<div>{JSON.stringify(data)}</div>);
        }}
      </Subscription>
    </ApolloProvider>
  });
};
```

**Comment by bmcmahen** [comment]
Hey @pashazz! For starters, I'd probably recommend not doing any grapqhl fetching within the `toasted-notes` callback itself. I'd fetch any relevant information within the component that is making the call to toasted-notes instead.

Having said that, you should be able to provide the `ApolloProvider` within the render callback like this:

```jsx
import client from './path/to/client';

const Notification = () => (
  <ApolloProvider client={client}>
    {/* subscriptions or queries go here */}
  </ApolloProvider>
);

toast.notify(<Notification />);
```

I haven't actually tried this so your mileage may vary. 

I think one solution I might build to better support context is to provide an optional `<ToastedNotesProvider />` component which effectively provides a target for toast notifications to render into.

**Comment by pashazz** [comment]
Thank you for your help. I've just done my GraphQL fetching that way:

```
    const observable = client.subscribe({
      query: TaskInfoUpdateDocument,
      variables: {
        ref: taskId,
      }
    });
    observable.subscribe(({data}) => {
      const func = async () => {
        client.cache.writeFragment<TaskDataType>({
          fragment: TaskFragmentFragmentDoc,
          id: dataIdFromObject(data.task),
          data: await getTask(data.task),
          fragmentName: "TaskFragment",
        });
      };
      func();
```

**Comment by bmcmahen** [comment]
Yeah, I don't think the `readCache` performs updates when your cache actually changes.

The `subscribe` function should be called when your cache changes, I think. So you could update component state when the subscribe callback is called, and then have your components update in response to those state changes. 

Buuut, I'm honestly not sure. I've probably try the `<ApolloProvider />` approach that I mentioned above. I can maybe hack together an example later tonight if I have some time.

**Comment by pashazz** [comment]
What do you mean exactly?  I intend to update the UI of notification when my subscription updates.  I use cache for that.  If only I had a method to trigger an update of my notification UI from subscribe callback,  It'd resolve my problems.  I can probably call notify from subscribe callback but then Id have had many notification.  Now if I could call notify that I replace an existing notification using some ID that I can provide,  that'd be great

**Comment by iampava** [comment]
Hey! Any chance we'll get first-hand Context support? Without the need to wrap the component? 

I'd be more than happy to contribute to this with a PR.

**Comment by victorocna** [comment]
I second this. First-hand context support would be awesome. It's the missing piece for me since I use this package with some translatable messages.

**Comment by bmcmahen** [comment]
hey all, I don't really have time to work on this currently. Happy to accept a PR 👍

**Comment by victorocna** [comment]
Hello @bmcmahen! I will have a look at it in the next couple of days. Cheers 👍

**Comment by victorocna** [comment]
Hello everyone. I had a look at it last night and have some questions. As far as I can tell, the context is missing because the `ToastManager` component is not placed in the app, like `<div><ToastManager /></div>`, but created at the root level [like so](https://github.com/bmcmahen/toasted-notes/blob/master/src/Alert/index.ts#L3). Would you like a PR that will add a parent component for toasts that can be placed in the app? How about breaking changes, should these be avoided?

**Comment by bmcmahen** [comment]
That makes sense to me. I imagine you'd want something like `ToastProvider` which can be placed somewhere in the react tree. If toast context is provided, it'd render using the toast provider... otherwise (to avoid breaking changes), you'd just render toasts as they currently are. I've actually implemented a similar system internally for work and it's worked well for us. 

```jsx
<StyledComponentContextProvider>
  <ToastProvider>
    <App />
  </ToastProvider>
</StyledComponentContextProvider>
```
